Add Swagger overview / usage documentation (#10844)
CDCgov/prime-reportstream: prime-router/docs/api/README.md

is provided in this repo (examples/generate-jwt-python). See the README in that directory for instructions, summarized
here:
1. Install python dependencies for generating JWT
2. Set variables on lines 28-30 in the script (`my_client_id`, `my_kid`, and `my_rsa_keypair_file`)
3. Run the script! The penultimate curl contains the JWT used to retrieve a bearer token. The final curl contains an
   example using that bearer token.

#### API Key (x-functions-key)

This method utilizes a simple API key and a client ID that is mapped to a header value and sent with every request.
It is created and stored in the Azure cloud per Azure function by ReportStream engineers and provided to the end user
during onboarding.  Important to note that this method is deprecated due to security concerns and should only be used in
edge cases where Okta or server-server auth is not possible. 

## Developer Notes

The openapi spec for ReportStream is generated from annotations in the code. This facilitates accurate, timely
updates and ameliorates an otherwise lengthy, manual, repetitive, and error-prone process.

### Swagger UI Customization and Maintenance

> Since the swagger files are generated from annotations in the code, any changes must be made to the annotations

Reviewer: ```suggestion
Since the OpenAPI spec files (in yaml format) are generated from annotations in the code, any changes must be made to the annotations
```

# Swagger UI

Swagger UI is a tool that allows you to visually interact with ReportStream APIs that have been defined using the OpenAPI
Specification (OAS). It provides a user-friendly interface that makes it easier to understand and explore capabilities of
the API without requiring external tools such as curl or Postman.

This directory contains all open api spec resources:

> 1. ./: specs in yaml format derived from other process (vs auto generated specs under ./generated)

Reviewer 2: probably remove any mention of these manually curated yml files since they do not show up in swagger and will be replaced with annotations.

2. ./swagger-ui: swagger ui artifacts (with customization - see custom and maintain notes below)
3. ./generated: auto generated api spec(s) from annotated kotlin source files

## Purpose

> The main purpose of Swagger UI is to:

Reviewer 2: ```suggestion
Swagger UI enables:
```

1. **Documentation:** generates documentation in a terse yet approachable format that serves as great reference
   material.

2. **API Visualization:** provides a structured interface that displays API endpoints, request parameters, response 
   structures, and example requests and responses in a structured, readable, and navigable format.

3. **API Testing:** facilitates direct interaction with the API. You can input parameters, make requests, and
   see the responses right in your browser with no external tools.

## Usage
With an API selected the UI will display a list of endpoints. Each endpoint can be expanded to display the
usage, parameters, and expected output; as well as an interface to make an API call.

### Accessing the Swagger UI Interface

The interface consists of static web assets that are compiled and served similarly to our front-end web application.
The UI is served and accessible on the same server as the reportstream web ui / front-end using a particular URL / path.
To access it, navigate to TLD for your chosen environment and append `/swagger-ui` to the end of the root url.

**Staging** https://staging.prime.cdc.gov/swagger-ui

> **Local** https://localhost:8080/swagger-ui

Reviewer 2: Does this link actually work?

Reviewer: > Does this link actually work?

the URL on staging is to be decided, since just like front end - report stream page webUI - is serve through Akamai, and there is DNS masquerade to be done, e.g. front end on staging is on URL: https://staging.reportstream.cdc.gov/

but FE site is served from Azure Blob container: xxxxxxxxxxxx (obfuscated)

and it is accessible from URL: https://xxxxxxxxxxx.z13.web.core.windows.net/ obfustcated

Akamai thing will be delegated to DevOp since that's not in our scope

Please note that while all environments can be selected in the swagger-ui (regardless of your chosen URL),
you may encounter cross-origin restrictions when trying to access one environment from outside its URL.

### Authentication

Swagger has built-in support for authentication; it will map user-provided credentials into the appropriate headers and
even handle obtaining and refreshing a bearer token using OAuth.

Regardless of chosen authentication method, the steps are similar:
Near the top-right of the interface you'll find an `Authorize` button which will bring up a menu with the authentication
methods available (described below). Each option will require some inputs.

> #### OAuth

Reviewer 2: I would call this Okta instead of OAuth? Unless it works for other OAuth cases? If so, then document somewhere that presently this is used for Okta auth

OAuth utilizes a `client ID` and a `secret` that is encoded as a JSON Web-Token. This token is submitted to the server to
request a bearer token which will be used to authenticate further request. The bearer token expires after 5 minutes so
it must be periodically refreshed using a separate API call.

Provide the client ID / secret and select a scope in the Authorize menu and Swagger will automate the above OAuth process.

#### Server-to-server

> Based on OAuth2, this method is based on best practices outlined in the SMART on FHIR. It follows a similar process to

Reviewer 2: We need documentation here on how to actually use this. A user will have to use the python script in the repo to get a token, then use the token here... kind of gross.

OAuth above: a client secret, ID, and scope is encoded as a JWT and sent to the server; which responds with a bearer
token the client can use to authenticate and authorize further requests.

#### API Key (x-functions-key)

> This method utilizes a simple API key and a client ID that is mapped to a header value and sent with every request. It

Reviewer 2: How does a user get this key/where does it come from?

is less secure, deprecated, and should only be used in edge cases where OAuth is not possible.

## Notes

### Swagger UI Customization and Maintenance

> This section is currently empty.

Reviewer 2: We should answer questions like `How are the swagger docs generated?` and anything else a dev might want or need to know to maintain our swagger docs

Reviewer 2: I would also mention somewhere that the swagger docs come from annotations and maybe examples or something? I'll leave it to your best judgement

OAuth above: a client secret, ID, and scope is encoded as a JWT and sent to the server; which responds with a bearer
token the client can use to authenticate and authorize further requests.

> The initial JWT must be created outside of the swagger environment. An example using a python script is provided in this

Reviewer 2: I think you jump in to eagerly here, I like how you started the Okta section with: 

`OKTA provides an implementation of OAuth2 for authentication.`

but here you assume the reader knows server-server is based on OAuth2 and that they understand what a "JWT" is and how the auth works?

repo (examples/generate-jwt-python). See the README instructions in that directory for instructions, summarized here:
1. Install python dependencies for generating JWT
2. Set variables on lines 28-30 in the script (`my_client_id`, `my_kid`, and `my_rsa_keypair_file`)
3. Run the script! The penultimate curl contains the JWT used to retrieve a bearer token. The final curl contains an
   example using that bearer token.

#### API Key (x-functions-key)

This method utilizes a simple API key and a client ID that is mapped to a header value and sent with every request. It

> is less secure, deprecated, and should only be used in edge cases where OAuth is not possible. This key is generated

Reviewer 2: ```suggestion
is less secure, deprecated, and should only be used in edge cases where Okta or server-server auth is not possible. This key is generated
```

> and stored in the Azure cloud and provided by reportstream to the end user.

Reviewer 2: ```suggestion
and stored in the Azure cloud per Azure function and provided by ReportStream to the end user.
```

Reviewer 2: Make sure everywhere we mention "ReportStream" it is spelled like this.

## Developer Notes

The openapi spec for reportstream is generated from annotations in the code. This facilitates accurate and timely

> updates and an otherwise ameliorates a lengthy, manual, repetitive, and error-prone process.

Reviewer 2: ```suggestion
updates and otherwise ameliorates a lengthy, manual, repetitive, and error-prone process.
```
